# Trabajo Final – Friendly SQL & Python para Analytics

## Análisis Exploratorio del Dataset Loan Approval

**Estudiante:** Diego Ching Pinto

**Objetivo:**

Realizar un análisis exploratorio de un conjunto de datos de solicitudes de préstamo utilizando Python, Plotly y DuckDB para obtener estadísticas descriptivas, visualizar la información y realizar consultas SQL sobre los datos.
**Fuente del dataset:** Loan Approval Prediction Dataset (Kaggle)

In [1]:
# Instalación de DuckDB
!pip install duckdb -q

# Librerías
import pandas as pd
import numpy as np
import duckdb

import plotly.express as px

In [2]:
from google.colab import files

uploaded = files.upload()

Saving Copy of loan.xlsx to Copy of loan.xlsx


In [3]:
df = pd.read_excel("Copy of loan.xlsx")

df.head()

,Loan_ID,Gender,Married,Dependents,Education,Self_Employed,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History,Property_Area,Loan_Status
0,LP001002,Male,No,0,Graduate,No,5849,0.0,NaN,360.0,1.0,Urban,Y
1,LP001003,Male,Yes,1,Graduate,No,4583,1508.0,128.0,360.0,1.0,Rural,N
2,LP001005,Male,Yes,0,Graduate,Yes,3000,0.0,66.0,360.0,1.0,Urban,Y
3,LP001006,Male,Yes,0,Not Graduate,No,2583,2358.0,120.0,360.0,1.0,Urban,Y
4,LP001008,Male,No,0,Graduate,No,6000,0.0,141.0,360.0,1.0,Urban,Y


## 1. Exploración inicial del dataset

En esta sección se realiza una revisión general del conjunto de datos para conocer su tamaño, estructura, nombres de columnas y tipos de datos. Esta exploración permite comprender la información disponible antes de realizar el análisis.

In [4]:
# Dimensiones del dataset
print(f"Cantidad de filas: {df.shape[0]}")
print(f"Cantidad de columnas: {df.shape[1]}")

Cantidad de filas: 614
Cantidad de columnas: 13


In [5]:
# Información general
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             601 non-null    object 
 2   Married            611 non-null    object 
 3   Dependents         599 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      582 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         592 non-null    float64
 9   Loan_Amount_Term   600 non-null    float64
 10  Credit_History     564 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


In [6]:
# Últimos registros
df.tail()

,Loan_ID,Gender,Married,Dependents,Education,Self_Employed,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History,Property_Area,Loan_Status
609,LP002978,Female,No,0,Graduate,No,2900,0.0,71.0,360.0,1.0,Rural,Y
610,LP002979,Male,Yes,3+,Graduate,No,4106,0.0,40.0,180.0,1.0,Rural,Y
611,LP002983,Male,Yes,1,Graduate,No,8072,240.0,253.0,360.0,1.0,Urban,Y
612,LP002984,Male,Yes,2,Graduate,No,7583,0.0,187.0,360.0,1.0,Urban,Y
613,LP002990,Female,No,0,Graduate,Yes,4583,0.0,133.0,360.0,0.0,Semiurban,N


In [7]:
# Nombres de las columnas
df.columns

Index(['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status'],
      dtype='object')

## 2. Calidad de los datos

En esta sección se revisan valores nulos, registros duplicados y posibles inconsistencias que puedan afectar el análisis.

In [8]:
# Valores nulos
df.isnull().sum()

,0
Loan_ID,0
Gender,13
Married,3
Dependents,15
Education,0
Self_Employed,32
ApplicantIncome,0
CoapplicantIncome,0
LoanAmount,22
Loan_Amount_Term,14


In [9]:
# Porcentaje de valores nulos
(df.isnull().sum()/len(df)*100).round(2)

,0
Loan_ID,0.00
Gender,2.12
Married,0.49
Dependents,2.44
Education,0.00
Self_Employed,5.21
ApplicantIncome,0.00
CoapplicantIncome,0.00
LoanAmount,3.58
Loan_Amount_Term,2.28


In [10]:
# Registros duplicados
df.duplicated().sum()

np.int64(0)

In [11]:
# Tipos de datos
df.dtypes

,0
Loan_ID,object
Gender,object
Married,object
Dependents,object
Education,object
Self_Employed,object
ApplicantIncome,int64
CoapplicantIncome,float64
LoanAmount,float64
Loan_Amount_Term,float64


## 3. Estadística descriptiva

En esta sección se presentan estadísticas descriptivas de las variables numéricas y categóricas para comprender mejor la distribución de los datos.

In [12]:
# Estadísticas de variables numéricas
df.describe()

,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History
count,614.000000,614.000000,592.000000,600.00000,564.000000
mean,5403.459283,1621.245798,146.412162,342.00000,0.842199
std,6109.041673,2926.248369,85.587325,65.12041,0.364878
min,150.000000,0.000000,9.000000,12.00000,0.000000
25%,2877.500000,0.000000,100.000000,360.00000,1.000000
50%,3812.500000,1188.500000,128.000000,360.00000,1.000000
75%,5795.000000,2297.250000,168.000000,360.00000,1.000000
max,81000.000000,41667.000000,700.000000,480.00000,1.000000


In [13]:
# Estadísticas de variables categóricas
df.describe(include="object")

,Loan_ID,Gender,Married,Dependents,Education,Self_Employed,Property_Area,Loan_Status
count,614,601,611,599,614,582,614,614
unique,614,2,2,4,2,2,3,2
top,LP002990,Male,Yes,0,Graduate,No,Semiurban,Y
freq,1,489,398,345,480,500,233,422


In [14]:
# Frecuencia de la variable objetivo
df["Loan_Status"].value_counts()

,count
Loan_Status,
Y,422
N,192


## 4. Visualizaciones

En esta sección se presentan algunas visualizaciones que permiten comprender mejor la distribución de los datos y las características de los solicitantes de crédito.

In [15]:
fig = px.bar(
    df["Education"].value_counts().reset_index(),
    x="Education",
    y="count",
    title="Cantidad de solicitantes por nivel educativo",
    labels={"count": "Cantidad"}
)

fig.show()

**Interpretación:**  
Se observa que la mayoría de los solicitantes de préstamo cuentan con educación de nivel *Graduate*, mientras que un porcentaje menor corresponde a personas con nivel *Not Graduate*.

In [16]:
fig = px.pie(
    df,
    names="Loan_Status",
    title="Estado de aprobación de préstamos",
    hole=0.5
)

fig.show()

**Interpretación:**  
La mayor parte de las solicitudes de préstamo fueron aprobadas (Loan_Status = Y), lo que representa aproximadamente el 69% del total de registros.

In [17]:
fig = px.histogram(
    df,
    x="ApplicantIncome",
    nbins=30,
    title="Distribución de ingresos del solicitante"
)

fig.show()

**Interpretación:**  
La distribución de los ingresos presenta una concentración en los valores bajos y algunos casos con ingresos significativamente altos, lo que evidencia la presencia de valores atípicos.

In [18]:
fig = px.box(
    df,
    y="LoanAmount",
    title="Distribución del monto del préstamo"
)

fig.show()

**Interpretación:**  
La mayoría de los montos de préstamo se concentran alrededor de un rango similar; sin embargo, existen algunos préstamos con montos considerablemente mayores, identificados como valores atípicos.

In [19]:
fig = px.scatter(
    df,
    x="ApplicantIncome",
    y="LoanAmount",
    color="Loan_Status",
    title="Relación entre ingreso y monto del préstamo"
)

fig.show()

**Interpretación:**  
Se observa una tendencia donde los solicitantes con mayores ingresos pueden acceder a préstamos de mayor monto, aunque la relación no es completamente lineal y presenta una alta dispersión.

## 5. Consultas SQL con DuckDB

En esta sección se realizan consultas SQL sobre el DataFrame utilizando DuckDB para obtener información relevante del conjunto de datos.

### Consulta 1. Visualización de los primeros registros

In [20]:
duckdb.sql("""
SELECT *
FROM df
LIMIT 10;
""").df()

,Loan_ID,Gender,Married,Dependents,Education,Self_Employed,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History,Property_Area,Loan_Status
0,LP001002,Male,No,0,Graduate,No,5849,0.0,NaN,360.0,1.0,Urban,Y
1,LP001003,Male,Yes,1,Graduate,No,4583,1508.0,128.0,360.0,1.0,Rural,N
2,LP001005,Male,Yes,0,Graduate,Yes,3000,0.0,66.0,360.0,1.0,Urban,Y
3,LP001006,Male,Yes,0,Not Graduate,No,2583,2358.0,120.0,360.0,1.0,Urban,Y
4,LP001008,Male,No,0,Graduate,No,6000,0.0,141.0,360.0,1.0,Urban,Y
5,LP001011,Male,Yes,2,Graduate,Yes,5417,4196.0,267.0,360.0,1.0,Urban,Y
6,LP001013,Male,Yes,0,Not Graduate,No,2333,1516.0,95.0,360.0,1.0,Urban,Y
7,LP001014,Male,Yes,3+,Graduate,No,3036,2504.0,158.0,360.0,0.0,Semiurban,N
8,LP001018,Male,Yes,2,Graduate,No,4006,1526.0,168.0,360.0,1.0,Urban,Y
9,LP001020,Male,Yes,1,Graduate,No,12841,10968.0,349.0,360.0,1.0,Semiurban,N


### Consulta 2. Solicitantes con ingresos mayores a 10000

In [21]:
duckdb.sql("""
SELECT Loan_ID,
       ApplicantIncome,
       LoanAmount,
       Loan_Status
FROM df
WHERE ApplicantIncome > 10000;
""").df()

,Loan_ID,ApplicantIncome,LoanAmount,Loan_Status
0,LP001020,12841,349.0,N
1,LP001100,12500,320.0,N
2,LP001186,11500,286.0,N
3,LP001233,10750,312.0,Y
4,LP001350,13650,NaN,Y
5,LP001369,11417,225.0,Y
6,LP001401,14583,185.0,Y
7,LP001422,10408,259.0,Y
8,LP001448,23803,370.0,Y
9,LP001451,10513,160.0,N


### Consulta 3. Cantidad de solicitantes por nivel educativo

In [22]:
duckdb.sql("""
SELECT Loan_ID,
       ApplicantIncome,
       LoanAmount,
       Loan_Status
FROM df
WHERE ApplicantIncome > 10000;
""").df()

,Loan_ID,ApplicantIncome,LoanAmount,Loan_Status
0,LP001020,12841,349.0,N
1,LP001100,12500,320.0,N
2,LP001186,11500,286.0,N
3,LP001233,10750,312.0,Y
4,LP001350,13650,NaN,Y
5,LP001369,11417,225.0,Y
6,LP001401,14583,185.0,Y
7,LP001422,10408,259.0,Y
8,LP001448,23803,370.0,Y
9,LP001451,10513,160.0,N


### Consulta 4. Solicitantes con mayores ingresos

In [23]:
duckdb.sql("""
SELECT Education,
       COUNT(*) AS Total_Solicitantes
FROM df
GROUP BY Education;
""").df()

,Education,Total_Solicitantes
0,Not Graduate,134
1,Graduate,480


### Consulta 5. Promedio de ingresos y préstamos según el estado de aprobación

In [24]:
duckdb.sql("""
SELECT Loan_ID,
       ApplicantIncome,
       LoanAmount
FROM df
ORDER BY ApplicantIncome DESC
LIMIT 10;
""").df()

,Loan_ID,ApplicantIncome,LoanAmount
0,LP002317,81000,360.0
1,LP002101,63337,490.0
2,LP001585,51763,700.0
3,LP001536,39999,600.0
4,LP001640,39147,120.0
5,LP002422,37719,152.0
6,LP001637,33846,260.0
7,LP001448,23803,370.0
8,LP002624,20833,480.0
9,LP001922,20667,NaN


In [25]:
duckdb.sql("""
SELECT Loan_Status,
       ROUND(AVG(ApplicantIncome),2) AS Promedio_Ingreso,
       ROUND(AVG(LoanAmount),2) AS Promedio_Prestamo
FROM df
GROUP BY Loan_Status;
""").df()

,Loan_Status,Promedio_Ingreso,Promedio_Prestamo
0,Y,5384.07,144.29
1,N,5446.08,151.22


# 6. Hallazgos y conclusiones

- El dataset contiene 614 registros y 13 variables relacionadas con solicitudes de préstamo.
- Se identificaron algunos valores nulos en determinadas variables, aunque no se encontraron registros duplicados.
- La mayoría de los solicitantes son graduados.
- Aproximadamente el 69% de las solicitudes fueron aprobadas y el 31% rechazadas.
- La distribución de los ingresos presenta algunos valores atípicos con ingresos considerablemente altos.
- El análisis mediante SQL permitió obtener información resumida sobre los solicitantes y los préstamos de forma sencilla y eficiente.